doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,414 @@
+# Writing Custom Functions
+
+This guide will describe how to create custom function classes in SymPy.
+Custom functions use the same mechanisms as the {ref}`functions <functions>`
+that are included with SymPy such as the common {ref}`elementary functions
+<elementary-functions>` like {class}`~.exp()` or {class}`~.sin`, {ref}`special
+functions <special-functions>` like {class}`~.gamma()` or {class}`~.Si()`, and
+{ref}`combinatorial functions <combinatorial-functions>` and {ref}`number
+theory functions <ntheory-module>` like {class}`~.factorial()` or
+{func}`~.primepi()`. Thus custom user functions are able to do the same things
+as functions defined by SymPy.
+
+This guide describes how to define functions that map a subset of
+$\mathbb{C}^n$ to $\mathbb{C}$. Functions that accept or return other kinds of
+objects should subclass another class, such as {class}`~.Boolean`,
+{class}`~.MatrixExpr`, {class}`~.Expr`, or {class}`~.Basic`. Much of what is
```

Comment:
Hopefully this will change in the future if we can integrate the kind system into Function (https://github.com/sympy/sympy/issues/4787, which might change other aspects of this guide as well).